runtime/memprof.c

```diff
@@ -119,257 +123,416 @@ static uintnat next_mt_generate_binom;
+  t->deleted = 1;
+  t->user_data = Val_unit;
+  t->block = Val_unit;
+  if (t_idx < trackst.delete) trackst.delete = t_idx;
+  CAMLassert(t->idx_ptr == NULL);
+}
+
+/* The return value is an exception or [Val_unit] iff [*t_idx] is set to
+   [Invalid_index]. In this case, the entry is deleted.
+   Otherwise, the return value is a [Some(...)] block. */
+static inline value run_callback_exn(uintnat *t_idx, value cb, value param) {
+  struct tracked* t = &trackst.entries[*t_idx];
   value res;
-  uintnat wosize = caml_current_callstack_size(callstack_size);
-  CAMLassert(!caml_memprof_suspended);
-  caml_memprof_suspended = 1; /* => no samples in the call stack. */
-  res = caml_alloc(wosize, 0);
-  caml_memprof_suspended = 0;
-  caml_current_callstack_write(res);
+  CAMLassert(!t->callback_running && t->idx_ptr == NULL);
+
+  t->callback_running = 1;
+  t->idx_ptr = t_idx;
+  res = caml_callback_exn(cb, param);
+   /* The call above can modify [*t_idx] and thus invalidate [t]. */
+  if (*t_idx == Invalid_index) {
+    /* Make sure this entry has not been removed by [caml_memprof_set] */
+    return Val_unit;
+  }
+  t = &trackst.entries[*t_idx];
+  t->idx_ptr = NULL;
+  t->callback_running = 0;
+  if (Is_exception_result(res) || res == Val_unit) {
+    /* Callback raised an exception or returned None or (), discard
+       this entry. */
+    mark_deleted(*t_idx);
+    *t_idx = Invalid_index;
+  }
   return res;
 }
 
-/**** Handling postponed sampled blocks. ****/
-/* When allocating in from C code, we cannot call the callback,
-   because the [caml_alloc_***] are guaranteed not to do so. These
-   functions make it possible to register a sampled block in a
-   todo-list so that the callback call is performed when possible. */
-/* Note: the shorter the delay is, the better, because the block is
-   linked to a root during the delay, so that the reachability
-   properties of the sampled block are artificially modified. */
-
-#define POSTPONED_DEFAULT_QUEUE_SIZE 128
-static struct postponed_block {
-  value block;
-  value callstack;
-  uintnat occurrences;
-  enum ml_alloc_kind kind;
-} default_postponed_queue[POSTPONED_DEFAULT_QUEUE_SIZE],
-  *postponed_queue = default_postponed_queue,
-  *postponed_queue_end = default_postponed_queue + POSTPONED_DEFAULT_QUEUE_SIZE,
-  *postponed_tl = default_postponed_queue, /* Pointer to next pop */
-  *postponed_hd = default_postponed_queue; /* Pointer to next push */
-
-static struct postponed_block* postponed_next(struct postponed_block* p)
+/* Run all the needed callbacks for a given entry.
+   In case of a thread context switch during a callback, this can be
+   called in a reetrant way.
+   The index of the entry may change. It is set to [Invalid_index] if
+   the entry is discarded.
+   Returns:
+   - An exception result if the callback raised an exception
+   - Val_long(0) == Val_unit == None otherwise
+ */
+static value run_entry_callbacks_exn(uintnat* t_idx)
```

Comment:
The interface of this function never points out that it is also responsible for advancing `trackst.callback`. I think this could be clarified in the comment. (I suggested the name `handle_entry_callbacks` to clarify that it not only runs the callbacks, but deals with the corresponding global state in the expected way, but I think the current name is also okay.)